### PR TITLE
[CR] Fix and simplify UTF8-detection

### PR DESCRIPTION
**Pull request by namarrgon**
fixes: #11941
`nl_langinfo(CODESET)` is [POSIX](http://man7.org/linux/man-pages/man3/nl_langinfo.3.html) and returns 'UTF-8' for all permutations of `{u,U}{t,T}{f,F}{,-}8` if the locale that is set with `LC_CTYPE` supports the utf8 encoding. This introduces a header that should be available on Linux/OSX but won't add new dependencies.
It should work on OSX the same way it does on Linux but I couldn't test it neither on OSX nor Windows.

I'm not sure what define to guard the header with since there seem to be three for windows: `_WIN32`,`__WIN32__` and `WINDOWS`. I just used the same ones that already guarded the utf8-test and limited it to non-windows curses-builds.

### CODE REVIEW

**Comment by drbig** [comment]
:+1: This is so much better. Thanks @namarrgon!

**Comment by DavidKeaton** [comment]
SO much cleaner. Thanks man. :-)

**Comment by vickio** [comment]
This doesn't seem to work on OS X. `nl_langinfo(CODESET)` returns `US-ASCII`, which fails the check, though the terminal encoding is set to UTF-8 and `setlocale` returns `en_US.UTF-8`.

**Comment by drbig** [comment]
@vickio I guess I'm too hopeful, but did you try to test it after explicit call to `setlocale(LC_ALL, "")`? 

Like with:

```C
#include <stdio.h>
#include <locale.h>
#include <langinfo.h>

int main() {
  setlocale(LC_ALL, "");
  printf("Reported: '%s'\n", nl_langinfo(CODESET));
}
```

It does at least work here (Linux) and for mixed locale settings...

...did I mention Unicode is a *bitch* on Unices?

**Comment by namarrgon** [comment]
@vickio Which OSX version is this on? Are you using OSX' "Terminal"? Can you post the output of `locale` and `locale -a`?

**Comment by vickio** [comment]
Testing with @drbig's snippet above works, it returns `UTF-8`, but if I build cataclysm from this PR, I get the Unicode warning on startup. I'm building with `LOCALIZE=0`, so maybe that's why?

This is OS X Yosemite 10.10 running in Terminal.

`locale`:

	LANG="en_US.UTF-8"
	LC_COLLATE="en_US.UTF-8"
	LC_CTYPE="en_US.UTF-8"
	LC_MESSAGES="en_US.UTF-8"
	LC_MONETARY="en_US.UTF-8"
	LC_NUMERIC="en_US.UTF-8"
	LC_TIME="en_US.UTF-8"
	LC_ALL=

`locale -a`

	af_ZA
	af_ZA.ISO8859-1
	af_ZA.ISO8859-15
	af_ZA.UTF-8
	am_ET
	am_ET.UTF-8
	be_BY
	be_BY.CP1131
	be_BY.CP1251
	be_BY.ISO8859-5
	be_BY.UTF-8
	bg_BG
	bg_BG.CP1251
	bg_BG.UTF-8
	ca_ES
	ca_ES.ISO8859-1
	ca_ES.ISO8859-15
	ca_ES.UTF-8
	cs_CZ
	cs_CZ.ISO8859-2
	cs_CZ.UTF-8
	da_DK
	da_DK.ISO8859-1
	da_DK.ISO8859-15
	da_DK.UTF-8
	de_AT
	de_AT.ISO8859-1
	de_AT.ISO8859-15
	de_AT.UTF-8
	de_CH
	de_CH.ISO8859-1
	de_CH.ISO8859-15
	de_CH.UTF-8
	de_DE
	de_DE.ISO8859-1
	de_DE.ISO8859-15
	de_DE.UTF-8
	el_GR
	el_GR.ISO8859-7
	el_GR.UTF-8
	en_AU
	en_AU.ISO8859-1
	en_AU.ISO8859-15
	en_AU.US-ASCII
	en_AU.UTF-8
	en_CA
	en_CA.ISO8859-1
	en_CA.ISO8859-15
	en_CA.US-ASCII
	en_CA.UTF-8
	en_GB
	en_GB.ISO8859-1
	en_GB.ISO8859-15
	en_GB.US-ASCII
	en_GB.UTF-8
	en_IE
	en_IE.UTF-8
	en_NZ
	en_NZ.ISO8859-1
	en_NZ.ISO8859-15
	en_NZ.US-ASCII
	en_NZ.UTF-8
	en_US
	en_US.ISO8859-1
	en_US.ISO8859-15
	en_US.US-ASCII
	en_US.UTF-8
	es_ES
	es_ES.ISO8859-1
	es_ES.ISO8859-15
	es_ES.UTF-8
	et_EE
	et_EE.ISO8859-15
	et_EE.UTF-8
	eu_ES
	eu_ES.ISO8859-1
	eu_ES.ISO8859-15
	eu_ES.UTF-8
	fi_FI
	fi_FI.ISO8859-1
	fi_FI.ISO8859-15
	fi_FI.UTF-8
	fr_BE
	fr_BE.ISO8859-1
	fr_BE.ISO8859-15
	fr_BE.UTF-8
	fr_CA
	fr_CA.ISO8859-1
	fr_CA.ISO8859-15
	fr_CA.UTF-8
	fr_CH
	fr_CH.ISO8859-1
	fr_CH.ISO8859-15
	fr_CH.UTF-8
	fr_FR
	fr_FR.ISO8859-1
	fr_FR.ISO8859-15
	fr_FR.UTF-8
	he_IL
	he_IL.UTF-8
	hi_IN.ISCII-DEV
	hr_HR
	hr_HR.ISO8859-2
	hr_HR.UTF-8
	hu_HU
	hu_HU.ISO8859-2
	hu_HU.UTF-8
	hy_AM
	hy_AM.ARMSCII-8
	hy_AM.UTF-8
	is_IS
	is_IS.ISO8859-1
	is_IS.ISO8859-15
	is_IS.UTF-8
	it_CH
	it_CH.ISO8859-1
	it_CH.ISO8859-15
	it_CH.UTF-8
	it_IT
	it_IT.ISO8859-1
	it_IT.ISO8859-15
	it_IT.UTF-8
	ja_JP
	ja_JP.eucJP
	ja_JP.SJIS
	ja_JP.UTF-8
	kk_KZ
	kk_KZ.PT154
	kk_KZ.UTF-8
	ko_KR
	ko_KR.CP949
	ko_KR.eucKR
	ko_KR.UTF-8
	lt_LT
	lt_LT.ISO8859-13
	lt_LT.ISO8859-4
	lt_LT.UTF-8
	nl_BE
	nl_BE.ISO8859-1
	nl_BE.ISO8859-15
	nl_BE.UTF-8
	nl_NL
	nl_NL.ISO8859-1
	nl_NL.ISO8859-15
	nl_NL.UTF-8
	no_NO
	no_NO.ISO8859-1
	no_NO.ISO8859-15
	no_NO.UTF-8
	pl_PL
	pl_PL.ISO8859-2
	pl_PL.UTF-8
	pt_BR
	pt_BR.ISO8859-1
	pt_BR.UTF-8
	pt_PT
	pt_PT.ISO8859-1
	pt_PT.ISO8859-15
	pt_PT.UTF-8
	ro_RO
	ro_RO.ISO8859-2
	ro_RO.UTF-8
	ru_RU
	ru_RU.CP1251
	ru_RU.CP866
	ru_RU.ISO8859-5
	ru_RU.KOI8-R
	ru_RU.UTF-8
	sk_SK
	sk_SK.ISO8859-2
	sk_SK.UTF-8
	sl_SI
	sl_SI.ISO8859-2
	sl_SI.UTF-8
	sr_YU
	sr_YU.ISO8859-2
	sr_YU.ISO8859-5
	sr_YU.UTF-8
	sv_SE
	sv_SE.ISO8859-1
	sv_SE.ISO8859-15
	sv_SE.UTF-8
	tr_TR
	tr_TR.ISO8859-9
	tr_TR.UTF-8
	uk_UA
	uk_UA.ISO8859-5
	uk_UA.KOI8-U
	uk_UA.UTF-8
	zh_CN
	zh_CN.eucCN
	zh_CN.GB18030
	zh_CN.GB2312
	zh_CN.GBK
	zh_CN.UTF-8
	zh_HK
	zh_HK.Big5HKSCS
	zh_HK.UTF-8
	zh_TW
	zh_TW.Big5
	zh_TW.UTF-8
	C
	POSIX

**Comment by drbig** [comment]
`LOCALIZE=0` does disable a lot of code, though I believe I've tried to put my code after the `setlocale()` call.

@namarrgon Can you please make sure that the check is called after the `setlocale(LC_ALL, "")` call? If that call doesn't happen with `LOCALIZE=0` then it should (it supposed to set a 'portable locale', whatever that *means*).

**Comment by DavidKeaton** [comment]
@drbig that means whatever the eight _million_ different permutations of a _supposed_ standard of POSIX are. :-P

**Comment by namarrgon** [comment]
Indeed, [`translations.cpp`](https://github.com/CleverRaven/Cataclysm-DDA/blob/d5a29cc8/src/translations.cpp#L68) calls `setlocale(LC_ALL, "")` only for LOCALIZE=1 and non-windows. We want to call it regardless of localization, in the worst case the program-internal locale stays the same. With f170e16 it is called for non-localized builds as well but it duplicates code, which is bad. I'm not sure how to best refactor set_language() to avoid this.

**Comment by drbig** [comment]
I would say `setlocale(LC_ALL, "")` is indeed localization-independent, if it works on our Windows build in a reasonable way (and that's not a highly set bar) then I'd prefer to see it before the proper localization stuff kicking in.

Anyone around doing Windows builds?

**Comment by namarrgon** [comment]
`setlocal(LC_ALL, "")` is now called from `main.cpp` and for both LOCALIZE=0 and =1 right before the call to `set_language()`, so functionality shouldn't have changed. I noticed a couple of bugs with the whole change-language-during-runtime thing but they aren't related to this PR and were already present.

It would be nice if a Windows-user could test removing the _WIN32 and WINDOWS guard around setlocale() to see whether that breaks the build: https://github.com/namarrgon/Cataclysm-DDA/blob/8e9766bb/src/main.cpp#L358-L367
MS' documentation suggests that it should work but a test wouldn't hurt …

I've tested this for a while now and it does seem to work as expected without opening up new bugs.

**Comment by drbig** [comment]
Anyone with some compilation environment to test this on Windows yet?

**Comment by i2amroy** [comment]
So I tested this in Code::Blocks and compiled with LOCALIZE=1 and it compiled just fine (once I had installed libintl and libiconv since they don't come with C::B by default) with or without that windows check. That said in both versions I was seeing %1$s type things showing up all over, and personally I don't know enough about localization code to understand if that was due to something here, if I was just missing a library, or if I had done something else wrong. So... yeah.

File: src/main.cpp [hunk 1]
```diff
@@ -354,6 +354,18 @@ int main(int argc, char *argv[])
...
+#if (defined _WIN32 || defined WINDOWS)
+    // Use the ANSI code page 1252 to work around some language output bugs.
+    if (setlocale(LC_ALL, ".1252") == NULL) {
+        DebugLog(D_WARNING, D_MAIN) << "Error while setlocale(LC_ALL, '.1252').";
+    }
+#else
+    if (setlocale(LC_ALL, "") == NULL) {
+        DebugLog(D_WARNING, D_MAIN) << "Error while setlocale(LC_ALL, '').";
+    }
+#endif
```
Comment (drbig) [inline src/main.cpp]: My proposal here would be to just call `setlocale(LC_ALL, "")` (with the `NULL` check as-is) without all the guards.

Then the `setlocale(LC_ALL, ".1252")` (also with the `NULL` check as-is) could be shoved back to translations.cpp and run only for `LOCALIZE=1` on Windows platforms. I'd let it be there due to that comment:

> Use the ANSI code page 1252 to work around some language output bugs.

No idea what 'bugs' there might be, but it seems better to stay on the 'safe(er) side' of things...

**Comment by namarrgon** [comment]
I'm still not sure why we use codepage 1252 for windows and what bugs it fixes (I couldn't find any obvious bugs when forcing it to the same utf8 locale) but git blame says that it was added with b12fbb61, perhaps @orzogc can shed some light into this?
If it turns out to be unnecessary, we can remove that any time in the future, for now I left that workaround in there, enabled setlocale(LC_ALL, "") for all other builds as @drbig suggested. I tested it on Linux and Windows (via codeblocks) and think that it's good to go in.

**Pull request merged**